# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/PContent.ipdl
# Commit: c0849928f934
# Full Hash: c0849928f934f0b5e931813eaadc23f0fb8521d7
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-04-23 09:52:48
# Regressor Bug: 1580565
# File Overlap Count: 2
# Description:
#   Bug 1580565 - Part 6: Add a unique ID to each BrowsingContextGroup, r=kmag
#   
#   This allows us to explicitly specify BrowsingContextGroups when synchronizing
#   them. A major advantage of this is that it means we can handle an attempt to
#   create a BrowsingContext with a parent which the content process is unaware of,
# ==============================================================================

diff -r 3d4f155096be -r c0849928f934 dom/ipc/PContent.ipdl
--- a/dom/ipc/PContent.ipdl	Wed Apr 22 15:48:15 2020 +0000
+++ b/dom/ipc/PContent.ipdl	Wed Apr 22 16:25:18 2020 +0000
@@ -832,7 +832,7 @@
 
     // Begin subscribing to a new BrowsingContextGroup, sending down the current
     // value for every individual BrowsingContext.
-    async RegisterBrowsingContextGroup(SyncedContextInitializer[] aInits);
+    async RegisterBrowsingContextGroup(uint64_t aGroupId, SyncedContextInitializer[] aInits);
 
 #if defined(MOZ_SANDBOX) && defined(MOZ_DEBUG) && defined(ENABLE_TESTS)
     // Initialize top-level actor for testing content process sandbox.
@@ -1569,32 +1569,23 @@
                     uint32_t flags);
 
     /**
-     * Sync the BrowsingContext with id 'aContextId' and name 'aName' to the
-     * parent, and attach it to the BrowsingContext 'aParentContext'. If
-     * 'aParentContext' is 'nullptr' the BrowsingContext is a root in the
-     * BrowsingContext tree. AttachBrowsingContext must only be called at most
-     * once for any child BrowsingContext, and only for BrowsingContexts where
-     * the parent and the child context contains their nsDocShell.
+     * Creates a new BrowsingContext, initialized with the values provided in
+     * `BrowsingContextInitializer`.
+     *
+     * This message may only be sent to the parent in limited situations. If the
+     * new BrowsingContext has a parent window, it must be owned by the
+     * embedding process, otherwise it must be owned by the opener, if set.
      */
-    async AttachBrowsingContext(BrowsingContextInitializer aInit);
+    async CreateBrowsingContext(uint64_t aGroupId, BrowsingContextInitializer aInit);
 
     /**
-     * Remove the synced BrowsingContext with id 'aContextId' from the parent.
-     * DetachBrowsingContext is only needed to be called once for any
-     * BrowsingContext, since detaching a node in the BrowsingContext detaches
-     * the entire sub-tree rooted at that node. Calling DetachBrowsingContext
-     * with an already detached BrowsingContext effectively does nothing. Note
-     * that it is not an error to call DetachBrowsingContext on a
-     * BrowsingContext belonging to an already detached subtree.
-     *
-     * As the passed-in context is allowed to already be detached, it is passed
-     * by id, rather than using BrowsingContext's serializer.
-     *
-     * Callers should keep the BrowsingContext alive until this async request is
-     * resolved or rejected, in order to ensure that in-flight messages still
-     * have valid targets.
+     * Discards the passed-in BrowsingContext. If the BrowsingContext has
+     * already been discarded, this message does nothing.
+     * The response promise is fulfilled when the process has flagged the
+     * BrowsingContext as discarded.
      */
-    async DetachBrowsingContext(uint64_t aContextId) returns (bool unused);
+    async DiscardBrowsingContext(MaybeDiscardedBrowsingContext aContext)
+            returns (bool unused);
 
     async WindowClose(MaybeDiscardedBrowsingContext aContext,
                       bool aTrustedCaller);